### PR TITLE
Add find a rep search controller

### PR DESCRIPTION
**Pull request by jvcAdHoc**
## Summary

- This pr adds the Find a Rep search endpoint
- It allows for searching based on type: organization, attorney, or claims agent
- It searches for the type within the max distance of a specified lat + long
- It can optionally fuzzy search on name
- It supports sorting on distance and relevenat name fields
- It has pagination

## Related issue(s)
- department-of-veterans-affairs/va.gov-team#66847

## Testing done

- I filled my local db with orgs and reps and did lots of postman requests:
  - moving the search lat + long around to test edge cases
  - testing sorting
  - testing invalid params
  - making sure pagination works with the params
- There has been extensive testing already on the distance and fuzzy name searching from the prs that added those.

## What areas of the site does it impact?
This isn't used by vets-website yet, but will feed the Find a Rep search results. It is also guarded by a feature flag.

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

### CODE REVIEW

**Comment by jvcAdHoc** [comment]
eod note: need to refactor controller and add specs

File: modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+      DEFAULT_PAGE = 1
+      DEFAULT_PER_PAGE = 10
+      DEFAULT_SORT = 'distance_asc'
+      PERMITTED_TYPES = %w[attorney claims_agent organization].freeze
+
+      PERMITTED_ORGANIZATION_SORTS = %w[distance_asc distance_desc name_asc name_desc].freeze
+      PERMITTED_REPRESENTATIVE_SORTS = %w[distance_asc distance_desc first_name_asc first_name_desc last_name_asc
+                                          last_name_desc].freeze
+      ORGANIZATION_SORT_MAP = {
```
Comment (jvcAdHoc) [inline modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb:21]: this isn't used

File: modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+          query_results = if search_params[:name]
+                            base_representatives_query.find_with_name_similar_to(search_params[:name])
+                          else
+                            base_representatives_query
+                          end
+          collection = Common::Collection.new(Veteran::Service::Representative, data: query_results)
+          resource = collection.paginate(**pagination_params)
+          render json: resource.data,
+                 serializer: CollectionSerializer,
+                 each_serializer: Veteran::Service::RepresentativeSerializer,
```
Comment (jvcAdHoc) [inline modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb:52]: need to make the serializers

File: modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      def index
+        if @search_model_type == 'organization'
+          query_results = if search_params[:name]
+                            base_organizations_query.find_with_name_similar_to(search_params[:name])
+                          else
+                            base_organizations_query
+                          end
+          collection = Common::Collection.new(Veteran::Service::Organization, data: query_results)
+          resource = collection.paginate(**pagination_params)
+          render json: resource.data,
```
Comment (holdenhinkle) [inline modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb:36]: I'm pretty sure that calling `render` twice will throw an error.

Are you wanting to send both `resource.data` and `query_results` as `JSON`? If so, they'll need to be combined.

Comment (jvcAdHoc) [reply modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb:36]: Good catch. That's a testing artifact that got uncommented before I pushed.

File: modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+      def serializer_klass
+        @serializer_klass ||= if organization_search?
+                                'Veteran::Accreditation::OrganizationSerializer'.constantize
+                              else
+                                'Veteran::Accreditation::RepresentativeSerializer'.constantize
+                              end
+      end
+
+      def base_query
+        @base_query ||= model_klass.find_within_max_distance(search_params[:long],
```
Comment (holdenhinkle) [inline modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb:55]: Do we need to set this `@base_query` ivar? It looks like `#base_query` is only called once.

File: modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+                                                             search_params[:lat]).order(sort_query_string)
+      end
+
+      def model_adjusted_query
+        @model_adjusted_query ||= if organization_search?
```
Comment (holdenhinkle) [inline modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb:60]: Do we need to set this `@model_adjusted_query` ivar? It looks like `#model_adjusted_query` is only called once.

File: modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+                                    base_query.select("veteran_organizations.*, #{distance_query_string}")
+                                  else
+                                    base_query.select("veteran_representatives.*, #{distance_query_string}").where(
+                                      '? = ANY(user_types) ', search_params[:type]
+                                    )
+                                  end
+      end
+
+      def accreditation_query
+        @accreditation_query ||= if search_params[:name]
```
Comment (holdenhinkle) [inline modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb:70]: Do we need to set this `@accreditation_query` ivar? It looks like `#accreditation_query` is only called once.

**Review by holdenhinkle — COMMENTED** [review]
This looks great. Just a few questions about setting instance variables. The same questions might apple to `#distance_query_string` + `@distance_query_string` and `#sort_query_string` + `@sort_query_string`.

**Comment by jvcAdHoc** [comment]
@holdenhinkle I removed all of the unnecessary memoization

File: modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb [hunk 7]
```diff
@@ -0,0 +1,160 @@
+# frozen_string_literal: true
+
+module Veteran
+  module V0
+    class AccreditedRepresentativesController < ApplicationController
+      skip_before_action :authenticate
```
Comment (jvcAdHoc) [inline modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb:6]: We don't want to require users to log in to search

File: modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+      before_action :verify_type
+      before_action :verify_sort
+      before_action :verify_long
+      before_action :verify_lat
+
+      DEFAULT_PAGE = 1
+      DEFAULT_PER_PAGE = 10
+      DEFAULT_SORT = 'distance_asc'
+
+      PERMITTED_TYPES = %w[attorney claim_agents organization].freeze
```
Comment (jvcAdHoc) [inline modules/veteran/app/controllers/veteran/v0/accredited_representatives_controller.rb:17]: @cosu419 note that the backend uses `claim_agents`, not `claims_agent`

**Review by holdenhinkle — APPROVED** [review]
This looks great, Jon!

**Review by ericboehs — APPROVED** [review]
